src/onnx/parse_if.cpp

```diff
@@ -65,13 +65,96 @@ struct parse_if : op_parser<parse_if>
 
         auto then_out_shapes = then_mdl->get_output_shapes();
         auto else_out_shapes = else_mdl->get_output_shapes();
-        if(not std::equal(then_out_shapes.begin(),
-                          then_out_shapes.end(),
-                          else_out_shapes.begin(),
-                          else_out_shapes.end()))
+
+        assert(then_out_shapes.size() == else_out_shapes.size());
+
+        // Must have the same type for both if/else blocks by onnx spec
+        // Add exception for empty constant scalars
+        if(then_out_shapes.at(0).type() != else_out_shapes.at(0).type() &&
+           (then_out_shapes.at(0).elements() > 0) && (else_out_shapes.at(0).elements() > 0))
         {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_grahps must have same output type! " +
+                           then_out_shapes.at(0).type_string() + " vs " +
+                           else_out_shapes.at(0).type_string());
+        }
+
+        if(not then_out_shapes.at(0).dynamic() && not else_out_shapes.at(0).dynamic())
+        {
+            if(then_out_shapes.at(0).scalar() && not else_out_shapes.at(0).scalar())
+            {
+                auto convert_ins = std::prev(then_mdl->end());
+                if(then_out_shapes.at(0).type() != else_out_shapes.at(0).type() &&
+                   then_out_shapes.at(0).elements() < 1)
+                {
+                    convert_ins = then_mdl->insert_instruction(
+                        convert_ins,
+                        migraphx::make_op("convert",
+                                          {{"target_type", else_out_shapes.at(0).type()}}),
+                        convert_ins->inputs().back());
+                    // then_mdl->replace_return({convert_ins});
+                }
```

Comment:
Is this for the case where the output types are not the same but one output shape is scalar? Why would a scalar have the wrong type?

Reply:
Empty shapes not being parsed in as the intended type. That may be an erroneous assumption since my other changes in the parse_constant PR seem to identify that shouldn't be the case

Reply:
More on this, its not just empty shapes but empty constants. We can get the type information but not anything about the shape, as it would register as an empty. Your fix for an empty literal will capture it correctly as having a shape::type {} {} 

In Netron it even shows the type, but has no information about the size.